src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,460 @@
+                log.info("Migrated solution build plan for exercise {} in {}ms", solutionParticipation.getProgrammingExercise().getId(), System.currentTimeMillis() - startMs);
+            });
+
+            Page<ProgrammingExerciseStudentParticipation> studentParticipationPage = ciMigrationService.orElseThrow()
+                    .getPageableStudentParticipations(programmingExerciseStudentParticipationRepository, pageable);
+            log.info("Found {} student programming exercises to migrate in batch", studentParticipationPage.getTotalElements());
+            studentParticipationPage.map(studentParticipation -> {
+                if (!ciMigrationService.orElseThrow().buildPlanExists(studentParticipation.getProgrammingExercise().getProjectKey(), studentParticipation.getBuildPlanId())) {
+                    log.warn("Build plan {} does not exist for exercise {}", studentParticipation.getBuildPlanId(), studentParticipation.getProgrammingExercise().getId());
+                    log.info("Skipping migration for exercise {} and setting build_plan_id to null", studentParticipation.getProgrammingExercise().getId());
+                    studentParticipation.setBuildPlanId(null);
+                    programmingExerciseStudentParticipationRepository.save(studentParticipation);
+                    return null;
+                }
+                return studentParticipation;
+            }).filter(Objects::nonNull).forEach(studentParticipation -> {
+                var startMs = System.currentTimeMillis();
+                List<AuxiliaryRepository> auxiliaryRepositories = new ArrayList<>();
+                if (ciMigrationService.orElseThrow().supportsAuxiliaryRepositories()) {
+                    auxiliaryRepositories = auxiliaryRepositoryRepository.findByExerciseId(studentParticipation.getProgrammingExercise().getId());
+                }
+                migrateStudentBuildPlan(studentParticipation.getProgrammingExercise(), studentParticipation, auxiliaryRepositories, errorMap);
+                log.info("Migrated student build plan for exercise {} in {}ms", studentParticipation.getProgrammingExercise().getId(), System.currentTimeMillis() - startMs);
+            });
+
+            log.info("Migrated {} / {} programming exercises in current thread", (Math.min(exerciseCount, batch + 1)), exerciseCount);
+            log.info("Migrated batch in {}ms", System.currentTimeMillis() - batchStart);
+        }
+    }
+
+    /**
+     * Migrates a single test build plan.
+     *
+     * @param exercise The exercise to migrate
+     * @param errorMap The error map to write errors to
+     */
+    private void migrateTestBuildPlan(ProgrammingExercise exercise, Map<ProgrammingExercise, Boolean> errorMap) {
+        if (exercise.getTestRepositoryUrl() == null) {
+            /*
+             * when the test repository url is null, we don't have to migrate the test build plan, saving multiple API calls
```

Comment:
Can you also explain why we don't have to do this?

Reply:
wdym?

Reply:
Tbh, I don't know what I meant any more 😕 I'll just resolve this